specs/networking/p2p-interface.md

```diff
@@ -0,0 +1,743 @@
+
+Two multiplexers are commonplace in libp2p implementations: [mplex](https://github.com/libp2p/specs/tree/master/mplex) and [yamux](https://github.com/hashicorp/yamux/blob/master/spec.md). Their protocol IDs are, respectively: `/mplex/6.7.0` and `/yamux/1.0.0`.
+
+Clients MUST support [mplex](https://github.com/libp2p/specs/tree/master/mplex) and MAY support [yamux](https://github.com/hashicorp/yamux/blob/master/spec.md). If both are supported by the client, yamux must take precedence during negotiation. See the Rationale section of this document for tradeoffs.
+
+# ETH2 network interaction domains
+
+## Constants
+
+This section outlines constants that are used in this spec.
+
+- `RQRP_MAX_SIZE`: The max size of uncompressed req/resp messages that clients will allow.
+ Value: TBD
+- `GOSSIP_MAX_SIZE`: The max size of uncompressed gossip messages
+ Value: 1MB (estimated from expected largest uncompressed block size).
+- `SHARD_SUBNET_COUNT`: The number of shard subnets used in the gossipsub protocol.
+ Value: TBD
+
+## The gossip domain: gossipsub
+
+Clients MUST support the [gossipsub](https://github.com/libp2p/specs/tree/master/pubsub/gossipsub) libp2p protocol.
+
+**Protocol ID:** `/meshsub/1.0.0`
+
+**Gossipsub Parameters**
+
+*Note: Parameters listed here are subject to a large-scale network feasibility study.*
+
+The following gossipsub parameters will be used:
+
+- `D` (topic stable mesh target count): 6
+- `D_low` (topic stable mesh low watermark): 4
+- `D_high` (topic stable mesh high watermark): 12
+- `D_lazy` (gossip target): 6
+- `fanout_ttl` (ttl for fanout maps for topics we are not subscribed to but have published to, seconds): 60
+- `gossip_advertise` (number of windows to gossip about): 3
+- `gossip_history` (number of heartbeat intervals to retain message IDs): 5
+- `heartbeat_interval` (frequency of heartbeat, seconds): 1
+
+### Topics
+
+Topics are plain UTF-8 strings, and are encoded on the wire as determined by protobuf (gossipsub messages are enveloped in protobuf messages).
+
+Topic strings have form: `/eth2/TopicName/TopicEncoding`. This defines both the type of data being sent on the topic and how the data field of the message is encoded. (Further details can be found in [Messages](#Messages)).
+
+There are two main topics used to propagate attestations and beacon blocks to all nodes on the network. Their `TopicName`'s are:
+
+- `beacon_block` - This topic is used solely for propagating new beacon blocks to all nodes on the networks. Blocks are sent in their entirety. Clients who receive a block on this topic MUST validate the block proposer signature before forwarding it across the network.
+- `beacon_attestation` - This topic is used to propagate aggregated attestations (in their entirety) to subscribing nodes (typically block proposers) to be included in future blocks. Similarly to beacon blocks, clients will be expected to perform some sort of validation before forwarding, but the precise mechanism is still TBD.
```

Comment:
what validation to be specific? signature? slot?

Reply:
Im thinking a signature validation but only require nodes to randomly perform to reduce overhead but still provide heard immunity from bad attestations. Need to consider this more before making a concrete proposal

Reply:
a reasonable start is not to propagate orphans (missing parent block) - once you're there, it seems pretty reasonable to require sig validation also. the concern is that constructing the state for the attestation is too costly?

Reply:
> a reasonable start is not to propagate orphans (missing parent block)

true. worth adding this IMO with a note on further validation.

> the concern is that constructing the state for the attestation is too costly

that plus the signature validation on all of the unaggregated (and aggregated) attestations might reduce our ability to gossip as quickly as necessary. Thinking the random validation with heard immunity is the way to go but need to do some modeling.

Reply:
a problem with random validations is that you can no longer disconnect peers that send you junk / police that they're doing their job

Reply:
https://github.com/ethereum/eth2.0-specs/issues/1332